components/eventing-controller/controllers/subscriptionv2/jetstream/reconciler.go

```diff
@@ -197,81 +184,121 @@ func (r *Reconciler) enqueueReconciliationForSubscriptions(subs []eventingv1alph
-	// check if the conditions are missing or changed
-	if !eventingv1alpha2.ConditionsEquals(sub.Status.Conditions, desiredConditions) {
-		sub.Status.Conditions = desiredConditions
-		updateStatus = true
-	}
-
-	// Update the status only if something needs to be updated
-	if updateStatus || readyStatusChanged {
-		if updateErr := r.Client.Status().Update(ctx, sub, &client.UpdateOptions{}); updateErr != nil {
-			events.Warn(r.recorder, sub, events.ReasonUpdateFailed, "Update Subscription status failed %s", sub.Name)
-			return pkgerrors.MakeError(errFailedToUpdateStatus, updateErr)
-		}
-		events.Normal(r.recorder, sub, events.ReasonUpdate, "Update Subscription status succeeded %s", sub.Name)
-	}
-	return nil
-}
-
 // handleSubscriptionDeletion deletes the JetStream subscription and removes its finalizer if it is set.
 func (r *Reconciler) handleSubscriptionDeletion(ctx context.Context,
-	subscription *eventingv1alpha2.Subscription, log *zap.SugaredLogger) error {
+	subscription *eventingv1alpha2.Subscription, log *zap.SugaredLogger) (ctrl.Result, error) {
+	// delete the JetStream subscription/consumer
 	if utils.ContainsString(subscription.ObjectMeta.Finalizers, eventingv1alpha2.Finalizer) {
 		if err := r.Backend.DeleteSubscription(subscription); err != nil {
 			// if failed to delete the external dependency here, return with error
 			// so that it can be retried
-			return pkgerrors.MakeError(errFailedToDeleteSub, err)
+			if syncErr := r.syncSubscriptionStatus(ctx, subscription, err, log); syncErr != nil {
+				return ctrl.Result{}, syncErr
+			}
+			return ctrl.Result{}, pkgerrors.MakeError(errFailedToDeleteSub, err)
 		}
 
-		// remove our finalizer from the list and update it.
+		// remove the eventing finalizer from the list and update the subscription.
 		subscription.ObjectMeta.Finalizers = utils.RemoveString(subscription.ObjectMeta.Finalizers,
 			eventingv1alpha2.Finalizer)
-		if err := r.Client.Update(ctx, subscription); err != nil {
-			events.Warn(r.recorder, subscription, events.ReasonUpdateFailed, "Update Subscription failed %s", subscription.Name)
-			return pkgerrors.MakeError(errFailedToRemoveFinalizer, err)
+		return ctrl.Result{}, r.syncSubscriptionStatus(ctx, subscription, nil, log)
+	}
+	return ctrl.Result{}, nil
+}
+
+// syncSubscriptionStatus syncs Subscription status and updates the k8s subscription.
+func (r *Reconciler) syncSubscriptionStatus(ctx context.Context,
+	desiredSubscription *eventingv1alpha2.Subscription, err error, log *zap.SugaredLogger) error {
+	// set ready state
+	desiredSubscription.Status.Ready = err == nil
+
+	// compile the desired conditions
+	desiredSubscription.Status.Conditions = eventingv1alpha2.GetSubscriptionActiveCondition(desiredSubscription, err)
+
+	// Update the subscription
+	return r.updateSubscription(ctx, desiredSubscription, log)
+}
+
+// updateSubscription updates the subscription changes to k8s.
+func (r *Reconciler) updateSubscription(ctx context.Context,
```

Comment:
this function updates or should only update the status. I see we touch finalizers as well here. Can we seperate these responsibilities?

Reply:
good idea, updated